x-pack/auditbeat/module/system/login/utmp.go

```diff
@@ -0,0 +1,571 @@
+	}
+}
+
+// UtmpFileReader can read a UTMP formatted file (usually /var/log/wtmp).
+type UtmpFileReader struct {
+	log           *logp.Logger
+	bucket        datastore.Bucket
+	filePattern   string
+	fileRecords   map[Inode]FileRecord
+	loginSessions map[string]LoginRecord
+}
+
+// NewUtmpFileReader creates and initializes a new UTMP file reader.
+func NewUtmpFileReader(log *logp.Logger, bucket datastore.Bucket, filePattern string) (*UtmpFileReader, error) {
+	r := &UtmpFileReader{
+		log:           log,
+		bucket:        bucket,
+		filePattern:   filePattern,
+		fileRecords:   make(map[Inode]FileRecord),
+		loginSessions: make(map[string]LoginRecord),
+	}
+
+	// Load state (file records, tty mapping) from disk
+	err := r.restoreStateFromDisk()
+	if err != nil {
+		return nil, errors.Wrap(err, "failed to restore state from disk")
+	}
+
+	return r, nil
+}
+
+// Close performs any cleanup tasks when the UTMP reader is done.
+func (r *UtmpFileReader) Close() error {
+	err := r.bucket.Close()
+	if err != nil {
+		return errors.Wrap(err, "error closing bucket")
+	}
+
+	return nil
+}
+
+// ReadNew returns any new UTMP entries in any files matching the configured pattern.
+func (r *UtmpFileReader) ReadNew() ([]LoginRecord, error) {
+	fileInfos, err := r.fileInfos()
+	defer r.deleteOldRecords(fileInfos)
+
+	var loginRecords []LoginRecord
+	for path, fileInfo := range fileInfos {
+		inode := Inode(fileInfo.Sys().(*syscall.Stat_t).Ino)
+
+		fileRecord, isKnownFile := r.fileRecords[inode]
+		var oldSize int64
+		if isKnownFile {
+			oldSize = fileRecord.Size
+		}
+		newSize := fileInfo.Size()
+
+		if newSize < oldSize {
+			// UTMP files are append-only and so this is weird. It might be a sign of
+			// a highly unlikely inode reuse - or of something more nefarious.
+			// Setting isKnownFile to false so we read the whole file from the beginning.
+			isKnownFile = false
+
+			r.log.Warnf("Unexpectedly, the file with inode %v (path=%v) is smaller than before - reading whole file.",
+				inode, path)
+		}
+
+		if !isKnownFile && newSize == 0 {
+			// Empty new file - save but don't read.
+			r.updateFileRecord(inode, newSize, nil)
+			continue
+		}
+
+		if !isKnownFile || newSize != oldSize {
+			r.log.Debugf("Reading file %v (inode=%v, oldSize=%v, newSize=%v)", path, inode, oldSize, newSize)
+
+			var utmpRecords []Utmp
+
+			// Once we start reading a file, we update the file record even if something fails -
+			// otherwise we will just keep trying to re-read very frequently forever.
+			defer r.updateFileRecord(inode, newSize, &utmpRecords)
+
+			if isKnownFile {
+				utmpRecords, err = r.readAfter(path, &fileRecord.LastUtmp)
+			} else {
+				utmpRecords, err = r.readAfter(path, nil)
+			}
+
+			if err != nil {
+				return nil, errors.Wrapf(err, "error reading file %v", path)
+			} else if len(utmpRecords) == 0 {
+				return nil, fmt.Errorf("unexpectedly, there are no new records in file %v", path)
```

Comment:
There's an `errors.Errorf` function that behaves similar to `errors.Wrap` in that it records stack information about where the error occurred. Also it might be useful for debugging purposes to include the number of bytes that were read (or the old and new sizes).